rclcpp/include/rclcpp/create_subscription.hpp

```diff
@@ -58,16 +64,49 @@ create_subscription(
   ),
   typename MessageMemoryStrategyT::SharedPtr msg_mem_strat = (
     MessageMemoryStrategyT::create_default()
-  )
+  ),
+  rclcpp::node_interfaces::NodeTimersInterface::SharedPtr node_timer_interface = nullptr
```

Comment:
Should there be validation for this value before it is used on line 99?

Reply:
Yes, that's a good point. If the user wants to use topic statistics and the interface is null then we should throw. I'll add the throw and document.

Reply:
This should be extracted from the `NodeT`. See line 72 for an example of another interface extracting it.

Reply:
I had tried that with compile errors that NodeT is a `rclcpp::node_interfaces::NodeTopicsInterface`, I'll double check.

Reply:
Being called from where? This one should be passing a `rclcpp::Node &`:

https://github.com/ros2/rclcpp/blob/0f0a4a8e397720ace85299eb30592e710b1b339b/rclcpp/include/rclcpp/node_impl.hpp#L96

Reply:
I thought the same given what was being passed, but maybe I'm interpreting the below incorrectly.

   Doing this
   ```
   create_subscription(
  NodeT && node,
  const std::string & topic_name,
  const rclcpp::QoS & qos,
  CallbackT && callback,
  const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> & options = (
    rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>()
  ),
  typename MessageMemoryStrategyT::SharedPtr msg_mem_strat = (
    MessageMemoryStrategyT::create_default()
  )
)
{
  using rclcpp::node_interfaces::get_node_timers_interface;
  auto node_timer_interface = get_node_timers_interface(std::forward<NodeT>(node));

  using rclcpp::node_interfaces::get_node_topics_interface;
  auto node_topics = get_node_topics_interface(std::forward<NodeT>(node));
  ```

  resulted in this

  ```
In file included from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/create_subscription.hpp:27:0,
                 from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node_impl.hpp:40,
                 from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node.hpp:1167,
                 from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/src/rclcpp/time_source.cpp:27:
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node_interfaces/get_node_timers_interface.hpp: In instantiation of ‘rclcpp::node_interfaces::NodeTimersInterface* rclcpp::node_interfaces::detail::get_node_timers_interface_from_pointer(NodeType) [with NodeType = std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface>*; typename std::enable_if<std::is_same<NodeType, std::shared_ptr<typename std::remove_pointer<_Tp>::type::element_type>*>::value, int>::type <anonymous> = 0]’:
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node_interfaces/get_node_timers_interface.hpp:143:56:   required from ‘rclcpp::node_interfaces::NodeTimersInterface* rclcpp::node_interfaces::get_node_timers_interface(NodeType&&) [with NodeType = std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface>&; typename std::enable_if<(! std::is_pointer<typename std::remove_reference< <template-parameter-1-1> >::type>::value), int>::type <anonymous> = 0]’
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/create_subscription.hpp:75:56:   required from ‘std::shared_ptr<SubscriptionT> rclcpp::create_subscription(NodeT&&, const string&, const rclcpp::QoS&, CallbackT&&, const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>&, typename MessageMemoryStrategyT::SharedPtr) [with MessageT = rosgraph_msgs::msg::Clock_<std::allocator<void> >; CallbackT = std::_Bind<void (rclcpp::TimeSource::*(rclcpp::TimeSource*, std::_Placeholder<1>))(std::shared_ptr<rosgraph_msgs::msg::Clock_<std::allocator<void> > >)>; AllocatorT = std::allocator<void>; CallbackMessageT = rosgraph_msgs::msg::Clock_<std::allocator<void> >; SubscriptionT = rclcpp::Subscription<rosgraph_msgs::msg::Clock_<std::allocator<void> >, std::allocator<void> >; MessageMemoryStrategyT = rclcpp::message_memory_strategy::MessageMemoryStrategy<rosgraph_msgs::msg::Clock_<std::allocator<void> >, std::allocator<void> >; NodeT = std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface>&; std::__cxx11::string = std::__cxx11::basic_string<char>; typename MessageMemoryStrategyT::SharedPtr = std::shared_ptr<rclcpp::message_memory_strategy::MessageMemoryStrategy<rosgraph_msgs::msg::Clock_<std::allocator<void> >, std::allocator<void> > >]’
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/src/rclcpp/time_source.cpp:232:3:   required from here
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node_interfaces/get_node_timers_interface.hpp:115:48: error: no matching function for call to ‘get_node_timers_interface_from_pointer(std::__shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface, (__gnu_cxx::_Lock_policy)2>::element_type*)’
   return get_node_timers_interface_from_pointer(node_shared_pointer->get());
   ```

Reply:
It's a specific instance, `time_source.cpp`. Likely it is trying to create the subscription on the cheap (without holding more interfaces).

So, one way to fix this would be to make `NodeTopicsInterface` depend on `NodeTimersInterface`, like it does with `NodeBaseInterface` right now.

We seem to be retracing some of my steps (which I'm not remembering until we happen on them again, lol).

I think we need to change the `NodeTopicsInterface` to have a `get_node_timers_interface()` equivalent to this:

https://github.com/ros2/rclcpp/blob/0f0a4a8e397720ace85299eb30592e710b1b339b/rclcpp/include/rclcpp/node_interfaces/node_topics_interface.hpp#L79-L82

And pass it when creating the `NodeTopics` instance:

https://github.com/ros2/rclcpp/blob/0f0a4a8e397720ace85299eb30592e710b1b339b/rclcpp/include/rclcpp/node_interfaces/node_topics.hpp#L42

Then, anything that has a `NodeTopicsInterface` also has the `NodeTimersInterface` needed to do this work. I think this is the least invasive way to do it, and it's what I did when I needed the node base interface in the topic specific part of rclcpp.

Sorry for the epic journey, but I'm fairly confident in this last approach. I think there's only a few places `NodeTopics` is instantiated:

- https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/node.cpp#L111
- https://github.com/ros2/rclcpp/blob/0f0a4a8e397720ace85299eb30592e710b1b339b/rclcpp_lifecycle/src/lifecycle_node.cpp#L70

And in both cases you have `NodeTimers` available right next to it.

Reply:
Thanks for the walkthrough, I was able to implement as you outlined.